### PR TITLE
update: pyquil dependency to v4

### PR DESCRIPTION
**Pull request by jselig-rigetti**
Left as a draft until the full release of [PyQuil v4](https://github.com/rigetti/pyquil/milestone/27). Changes between the RC used here and the eventual full release should be compatible, but leave this MR as a draft until the dependency can be pinned to a full >=4.0.0 release.

### CODE REVIEW

File: pyquil_for_azure_quantum/__init__.py [hunk 1]
```diff
@@ -182,7 +154,7 @@ def get_qvm() -> AzureQuantumComputer:
     Raises:
         KeyError: If required environment variables are not set.
     """
-    return AzureQuantumComputer(target="rigetti.sim.qvm", qpu_name="qvm")
+    return AzureQuantumComputer(target=RigettiTarget.QVM.value, qpu_name="qvm")
```
Comment (jselig-rigetti) [inline pyquil_for_azure_quantum/__init__.py:157]: Do we need to update https://github.com/microsoft/qdk-python/blob/main/azure-quantum/azure/quantum/target/rigetti/target.py#L151 as well?

File: pyquil_for_azure_quantum/__init__.py [hunk 2]
```diff
@@ -225,12 +196,13 @@ def __init__(self, *, target: str) -> None:
...
-        """Run the executable and wait for its results"""
-        return super().run(executable)
-
-    def execute(self, executable: AzureProgram, name: str = "pyquil-azure-job") -> AzureJob:  # type: ignore[override]
+    def execute(  # type: ignore[override]
+        self,
+        executable: AzureProgram,
+        memory_map: Optional[MemoryMap] = None,
+        name: str = "pyquil-azure-job",
+        **kwargs: Any,
```
Comment (BatmanAoD) [inline pyquil_for_azure_quantum/__init__.py:204]: What's `kwargs` for? I don't see it being used; am I just missing something?

File: pyquil_for_azure_quantum/__init__.py [hunk 3]
```diff
@@ -329,20 +305,34 @@ def run_batch(
...
-        combined_result = self.get_result(AzureJob(job, executable))
+        azure_job = AzureJob(job=job, executable=executable)
+        combined_result = self.get_result(azure_job)
         if num_params is None or num_params == 1:
             return [combined_result]
-        split_results = split(combined_result.readout_data["ro"], num_params)  # type: ignore
-        return [QAMExecutionResult(executable, {"ro": result}) for result in split_results]
-
 
-_RawData = Union[int, float, List[float]]
```
Comment (BatmanAoD) [inline pyquil_for_azure_quantum/__init__.py:348]: Is this no longer used?

**Review by BatmanAoD — COMMENTED** [review]
I think Marquess should probably review this if he has time, but I don't see any obvious problems.

File: pyquil_for_azure_quantum/__init__.py [hunk 4]
```diff
@@ -269,10 +237,18 @@ def get_result(self, execute_response: AzureJob) -> QAMExecutionResult:
         job = execute_response.job
         job.wait_until_completed()
         result = Result(job)
-        numpified = {k: array(v) for k, v in result.data_per_register.items()}
+
+        # pylint: disable-next=fixme
+        # TODO: as of https://github.com/microsoft/qdk-python/blob/4d6f7f75c8c7d8467f87936b1aaef449de1e0bf6/azure-quantum/azure/quantum/target/rigetti/result.py#L47
+        # both QVM and QC result shapes take the memory-map form as in the QVMRedustData.
```
Comment (MarquessV) [inline pyquil_for_azure_quantum/__init__.py:243]:
```suggestion
        # both QVM and QC result shapes take the memory-map form as in the QVMResultData.
```

File: pyquil_for_azure_quantum/__init__.py (hunk 1 above)
Comment (MarquessV) [reply pyquil_for_azure_quantum/__init__.py:156]: Sorry I'm missing it, but I'm not connecting the dots here. What should we update?

**Comment by MarquessV** [comment]
These changes look good, seems like the CI is still trying to run tests against Python 3.7, which is no longer supported, but once that is figured out LGTM.

File: .github/workflows/checks.yml [hunk 5]
```diff
@@ -84,7 +88,6 @@ jobs:
     strategy:
       matrix:
         python:
-          - 3.7
```
Comment (MarquessV) [inline .github/workflows/checks.yml:87]: Oh yeah, let's throw 3.11 in the mix while we are at it.

**Comment by MarquessV** [comment]
Reminder to [update main's branch protection settings](https://github.com/rigetti/pyquil-for-azure-quantum/settings/branch_protection_rules/) after this is merged. The 3.7 checks should be replaced with checks against 3.11.

File: pyquil_for_azure_quantum/__init__.py (hunk 3 above)
Comment (jselig-rigetti) [reply pyquil_for_azure_quantum/__init__.py:348]: Wasn't being used before or after the changes here 🤷

**Pull request merged**